### ISSUE TITLE
Performance/Concurrency

### ISSUE THREAD

**Issue by giuseppetotaro** [nasa-jpl/ASSESS#22]
We need to discuss if and how the prototype should be able to handle multiple requests at the same time.

**Comment by karimjp** [comment]
Hey guys,

Flask runs on top of a Werkzeug server.  The Flask application run method specifies an 'options' parameter that can be set for the werkzeug.serving.run_simple method which runs the Werkzeug builtin server.  I enabled and tested the  'threaded' and 'processes' parameters separately in the app.run Flask method, to test for any performance gains.  The result was that setting the parameter processes to an approximate number of the CPU cores available in the machine was the best option.  It is recommended in the Werkzeug documentation that for production it is better to use a full blown server like Apache for better performance.

In the following cases tested, the following URL pattern was generated 5 times for 5 asynchronous requests in a machine with 4 CPU cores:

URL Pattern: 
url = http://0.0.0.0:5000/predict | body = "test" + i
GET(url, params={ "text": body})

Case 1: Enabling 'processes' Option
  Code: 
  	app.run(processes=5, host="0.0.0.0", port=5000)
  Tests:
    Http Simultaneous Requests: 5
    Total seconds for each: 
    [23.822874069213867, 25.050405025482178, 25.123871088027954, 26.221457958221436, 26.562421083450317]

Case 2: Enabling 'threaded' Option
  Code:
    app.run(threaded=True,host="0.0.0.0", port=5000)
  Tests:
    Http Requests: 5
    Total seconds for each: 
    [84.36615014076233, 86.17623496055603, 89.14776611328125, 94.77311205863953, 95.25755500793457]

Case 3: No modifications
  Code:
    app.run(host="0.0.0.0", port=5000)
  Tests:
    Http Requests: 5
    Total seconds for each:
    [15.417919874191284, 31.09600806236267, 47.229403018951416, 63.22134304046631, 79.21947288513184]

Here's a link to the testing code I built:  https://github.com/karimjp/http-get-test/blob/master/get_test.py

Hope this helps start the conversation.


Here is a reference for more details:
http://flask.pocoo.org/docs/0.12/api/#flask.Flask.run
http://werkzeug.pocoo.org/docs/0.14/serving/#werkzeug.serving.run_simple

My Machine Specs:
root@4b9f7eaa0593:/ASSESS/webapp# lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             4
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 70
Model name:            Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
Stepping:              1
CPU MHz:               2194.558
BogoMIPS:              4440.51
Hypervisor vendor:     vertical
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              6144K
L4 cache:              131072K
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht pbe syscall nx pdpe1gb lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq dtes64 ds_cpl ssse3 sdbg fma cx16 xtpr pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm fsgsbase bmi1 avx2 bmi2 erms xsaveopt arat
